stan/math/prim/mat/functor/mpi_parallel_call.hpp

```diff
@@ -0,0 +1,488 @@
+                        local_x_r[i], local_x_i[i], 0);
+        local_f_out[i] = job_output.cols();
+
+        if (local_outputs_per_job == -1) {
+          local_outputs_per_job = job_output.rows();
+          local_output.conservativeResize(local_outputs_per_job,
+                                          Eigen::NoChange);
+        }
+
+        if (local_output.cols() < offset + local_f_out[i])
+          local_output.conservativeResize(Eigen::NoChange,
+                                          2 * (offset + local_f_out[i]));
+
+        local_output.block(0, offset, local_output.rows(), local_f_out[i])
+            = job_output;
+      }
+    } catch (const std::exception& e) {
+      local_ok = 0;
+    }
+
+    // during first execution we distribute the output sizes from
+    // local jobs to the root. This needs to be done for the number of
+    // outputs of each result and the number of outputs per job.
+    if (num_outputs_per_job_ == -1) {
+      // before we can cache the sizes locally we must ensure
+      // that no exception has been fired from any node. Hence,
+      // share the info about the current status across all nodes.
+      int cluster_status = 0;
+      boost::mpi::reduce(world_, local_ok, cluster_status, std::plus<int>(), 0);
+      bool all_ok = cluster_status == static_cast<int>(world_size_);
+      boost::mpi::broadcast(world_, all_ok, 0);
+      if (!all_ok) {
+        // err out on the root
+        if (rank_ == 0) {
+          throw std::domain_error("MPI error on first evaluation.");
+        } else {
+          // and ensure on the workers that they return into their
+          // listening state
+          return result_t();
+        }
+      }
+
+      // execution was ok everywhere such that we can now distribute the
+      // outputs per job as recorded on the root on the first execution
+      boost::mpi::broadcast(world_, local_outputs_per_job, 0);
+      num_outputs_per_job_ = local_outputs_per_job;
+
+      if (!cache_f_out::is_valid()) {
+        // ok, now all is fine everywhere, collect sizes and cache info
+        std::vector<int> world_f_out(num_jobs, 0);
+        boost::mpi::gatherv(world_, local_f_out.data(), num_local_jobs,
+                            world_f_out.data(), job_chunks, 0);
+        // on the root we now have all sizes from all childs. Copy
+        // over the local sizes to the world vector on each local
+        // node in order to cache this information locally.
+        std::copy(local_f_out.begin(), local_f_out.end(),
+                  world_f_out.begin() + start_job);
+        cache_f_out::store(world_f_out);
+      }
+    }
+
+    typename cache_f_out::cache_t& world_f_out = cache_f_out::data();
+
+    // check that cached sizes are the same as just collected from
+    // this evaluation
+    for (int i = 0; i < num_local_jobs; ++i) {
+      if (world_f_out[start_job + i] != local_f_out[i]) {
+        local_ok = 0;
```

Comment:
You have to continue to keep everyone synced. Assume I would quit immediately. That would mean that one of the workers won't reach the gather statement below... and this would mean that the root would wait forever for the worker which did a premature quit. Therefore, the code is setup to keep running and only transmit the success status at the very end.

Things happen in different processes and communication needs to stay synched. That does not make things easier...